File: src/NetMusicStore/Catalog.Domain/Albums/Album.cs
using Catalog.Domain.Aggregates.Songs;
using Catalog.Domain.Events;
using Catalog.Domain.Shared.ValueObjects;
using Shared.Domain;
using Shared.Domain.Bus.Event;
using Shared.Domain.ValueObjects;

namespace Catalog.Domain.Albums
{
    public class Album : AggregateRoot
    {
        public AggregateId<Album, string> Id { get; private set; }
        public AlbumType Type { get; private set; }
        public AlbumName Name { get; private set; }
        public AlbumDescription Description { get; private set; }
        public Author Author { get; private set; }
        public AuditValueObject Audit { get; private set; }
        public EAlbumStatus Status { get; private set; }

        private readonly List<AlbumTag> _tags;
        public IReadOnlyCollection<AlbumTag> Tags => _tags.AsReadOnly();
        
        private readonly List<Song> _songs;
        public IReadOnlyCollection<Song> Songs => _songs.AsReadOnly();


        private Album(AggregateId<Album, string> id,
                      AlbumType type,
                      AlbumName name,
                      AlbumDescription description,
                      Author author,
                      List<AlbumTag> tags)
        {
            Id = id;
            Name = name;
            Description = description;
            Author = author;
            Type = type;
            _tags = tags;
            _songs = new List<Song>();
            Audit =  AuditValueObject.Create(author.UserName);
            Status = EAlbumStatus.Pending;

            AddAlbumCreatedDomainEvent(id.Value, type.Description, name.Value, description.Value);
        }

        public void SetType(AlbumType type)
        {
            Type = type;
        }

        private void AddAlbumCreatedDomainEvent(string id, string type, string name, string description)
        {
            var createdEvent = new AlbumCreatedDomainEvent(id, type, name, description);

            AddDomainEvent(createdEvent);
        }

        public static Album Create(AlbumType type,
                                   AlbumName name,
                                   AlbumDescription description,
                                   Author author,
                                   List<AlbumTag> tags)
        {
            var id = AggregateId<Album, string>.From(Guid.NewGuid().ToString());

            return new Album(id, type, name, description, author, tags);
        }

        public void SendToLab()
        {
            if (Status != EAlbumStatus.Pending)
                StatusChangeException(EAlbumStatus.InLab);

            Status = EAlbumStatus.InLab;

            AddDomainEvent(new AlbumSentToLabDomainEvent(Id.Value));
        }

        public void SendToSales()
        {
            if (Status != EAlbumStatus.InLab)
                StatusChangeException(EAlbumStatus.InSales);

            Status = EAlbumStatus.InLab;

            AddDomainEvent(new AlbumSentToSalesDomainEvent(Id.Value));
        }

        public void Cancel()
        {
            if (Status != EAlbumStatus.Pending)
                StatusChangeException(EAlbumStatus.Canceled);

            Status = EAlbumStatus.Canceled;

            AddDomainEvent(new AlbumCanceledDomainEvent(Id.Value));
        }

        public void Close()
        {
            if (Status != EAlbumStatus.InSales)
                StatusChangeException(EAlbumStatus.Closed);

            Status = EAlbumStatus.Closed;

            AddDomainEvent(new AlbumClosedDomainEvent(Id.Value));
        }

        public void AddSong(string name, string description, double duration, List<string> tags, Author author)
        {
            if (ExistsSong(name))
                throw new DomainException($"Song named {name} exists.");

            _songs.Add(Song.Create(name, description, duration, _durationFormatter.Format(duration), tags, author));
        }

        public void RemoveSong(string name)
        {
            if (!ExistsSong(name))
                throw new DomainException($"Song named {name} does not exists.");

            var song = _songs.Find(p => p.Name.ToLower() == name.ToLower());

            _songs.Remove(song);
        }

        private bool ExistsSong(string name)
        {
            return _songs.Exists(p => p.Name.ToLower() == name.ToLower());
        }

        private void StatusChangeException(EAlbumStatus albumStatusToChange)
        {
            throw new AlbumDomainException($"Is not possible to change the order status from {Status.Name} to {albumStatusToChange.Name}.");
        }
    }
}

File: src/BackOffice/server/src/services/albums/domain/AlbumAggregate/Album.ts
import { AlbumId } from './AlbumId';
import { AlbumName } from './AlbumName';
import { AggregateRoot } from '../../../shared/domain/AggregateRoot';
import { AuditValueObject } from '../../../shared/domain/valueobjects';
import AlbumDomainException from './AlbumDomainException';
import Author from './Author';
import Song from './Song';
import SongDomainException from './SongDomainException';
import { AlbumDescription } from './AlbumDescription';
import { AlbumCreatedDomainEvent } from './events';
import { Tag } from './Tag';

export enum EAlbumStatus {
  Pending,
  InLab,
  Closed,
  Canceled
}

export const EAlbumStatusTranslate = {
  [EAlbumStatus.Pending]: 'Pending',
  [EAlbumStatus.InLab]: 'InLab',
  [EAlbumStatus.Closed]: 'Closed',
  [EAlbumStatus.Canceled]: 'Canceled'
};

class Album extends AggregateRoot {
  readonly id: AlbumId;
  readonly name: AlbumName;
  readonly description: AlbumDescription;
  readonly registerDate: Date;
  readonly author: Author;
  readonly tags: Tag[];
  readonly songs: Song[];
  audit: AuditValueObject;
  status: EAlbumStatus;

  private constructor(
    id: AlbumId,
    name: AlbumName,
    description: AlbumDescription,
    author: Author,
    tags: Tag[]
  ) {
    super();

    const registerDate = new Date();

    this.id = id;
    this.name = name;
    this.description = description;
    this.author = author;
    this.tags = tags;
    this.registerDate = registerDate;
    this.songs = [];
    this.audit = AuditValueObject.Create(this.author.authorId);
    this.status = EAlbumStatus.Pending;
  }

  public static Create(
    id: AlbumId,
    name: AlbumName,
    description: AlbumDescription,
    author: Author,
    tags: Tag[]
  ): Album {
    const album = new Album(id, name, description, author, tags);

    album.AddDomainEvent(
      new AlbumCreatedDomainEvent({
        id: album.id.value,
        name: album.name.value,
        description: album.description.value,
        registerDate: album.registerDate,
        author: album.author.authorId
      })
    );

    return album;
  }

  public Promote() {
    if (this.status !== EAlbumStatus.Pending)
      this.StatusChangeException(EAlbumStatus.InLab);

    this.status = EAlbumStatus.InLab;
    this.audit = this.audit.Update(this.author.authorId);
  }

  public Cancel() {
    if (this.status !== EAlbumStatus.Pending)
      this.StatusChangeException(EAlbumStatus.Canceled);

    this.status = EAlbumStatus.Canceled;
    this.audit = this.audit.Update(this.author.authorId);
  }

  public AddTag(tag: Tag): void {
    if (this.ExistsTag(tag))
      throw new AlbumDomainException(`Tag ${tag.value} exists`);

    this.tags.push(tag);
  }

  public RemoveTag(tag: Tag): void {
    if (!this.ExistsTag(tag))
      throw new AlbumDomainException(`Tag ${tag.value} does not exists`);

    this.tags.splice(
      this.tags.findIndex((p) => p.value === tag.value),
      1
    );
  }

  public AddSong(
    id: string,
    name: string,
    description: string,
    duration: number,
    durationFormatted: string,
    tags: string[],
    author: Author
  ): void {
    if (this.ExistsSong(name))
      throw new SongDomainException(`Song ${name} exists`);

    const song = Song.Create(
      id,
      name,
      description,
      duration,
      durationFormatted,
      tags,
      author
    );

    this.songs.push(song);
  }

  public RemoveSong(song: Song): void {
    const name = song.name;

    if (!this.ExistsSong(name))
      throw new AlbumDomainException(`Song ${name} does not exists`);

    this.songs.splice(
      this.songs.findIndex((p) => p.name.toLowerCase() === name.toLowerCase()),
      1
    );
  }

  private ExistsTag(tag: Tag): boolean {
    const tagFound = this.tags.find((t) => t.value === tag.value);

    return tagFound !== null ? true : false;
    
  }

  private ExistsSong(name: string): boolean {
    return this.songs.find((p) => p.name.toLowerCase() === name.toLowerCase())
      ? true
      : false;
  }

  private StatusChangeException(albumStatusToChange: EAlbumStatus) {
    throw new AlbumDomainException(
      `Is not possible to change the album status from ${
        EAlbumStatusTranslate[this.status]
      } to ${EAlbumStatusTranslate[albumStatusToChange]}`
    );
  }

  static fromPrimitives(plainData: {
    id: string;
    name: string;
    description: string;
    author: string;
    duration: string;
    tags: string[];
  }): Album {
    const TagsPrimitive = plainData.tags.map((tag) => new Tag(tag));

    return new Album(
      new AlbumId(plainData.id),
      new AlbumName(plainData.name),
      new AlbumDescription(plainData.description),
      Author.Create(plainData.author, '', ''),
      TagsPrimitive
    );
  }

  toPrimitives(): unknown {
    return {
      id: this.id,
      name: this.name,
      description: this.description,
      registerDate: this.registerDate,
      author: this.author,
      tags: this.tags
    };
  }
}

export default Album;
